src/differencer/jsonDiff.ts

```diff
@@ -0,0 +1,382 @@
+ */
+function parseGraph(graphDiffs: Array<object>): Array<NodeDiff> {
+  const typedDiffs: Array<NodeDiff> = [];
+  Object.keys(graphDiffs).forEach(key => {
+    if (key === JSON_DIFF_PATCH_ARRAY_KEY) {
+      //Ignoring the property added by jsondiffpath to indicate that graph is an array
+      return;
+    }
+    const diff = graphDiffs[key];
+    let typedDiff;
+    if (_.isArray(diff)) {
+      //Node is added/removed/moved
+      typedDiff = addNodeDiff(diff);
+    } else if (_.isObject(diff)) {
+      //properties in a node are added/removed/modified/moved
+      typedDiff = parseNodePropDiffs(diff[JSON_LD_KEY_ID], diff);
+    }
+    if (typedDiff != null) {
+      typedDiffs.push(typedDiff);
+    }
+  });
+  return typedDiffs;
+}
+
+/**
+ * Parse differences to the node properties
+ * @param nodeId ID of the node
+ * @param diff JSON node that has differences
+ *
+ * @returns Instance of NodeDiff for the node if there are differences
+ */
+function parseNodePropDiffs(nodeId: string, diff: object): NodeDiff {
+  let typedDiff = new NodeDiff(nodeId);
+  Object.keys(diff)
+    .filter(key => key !== JSON_LD_KEY_ID) //ignore node id
+    .forEach(key => {
+      const value = diff[key];
+
+      //check if the property of the node is an array
+      const isArray =
+        value.hasOwnProperty(JSON_DIFF_PATCH_ARRAY_KEY) &&
+        value[JSON_DIFF_PATCH_ARRAY_KEY] === JSON_DIFF_PATCH_ARRAY_VALUE;
+
+      if (isArray) {
+        //Ignore property added by jsondiffpath to indicate array property
+        Object.keys(value)
+          .filter(arrKey => arrKey !== JSON_DIFF_PATCH_ARRAY_KEY)
+          .forEach(arrKey => {
+            //Since this is flattened graph array value is either a string or a object with an id referencing other object in the graph
+            const arrValue = value[arrKey];
+            const diffType = getDiffType(arrValue);
+            ramlToolLogger.debug(
+              `Adding difference of node: ${typedDiff.id}, array property: ${key}, array key: ${arrKey}, difference type: ${DiffType[diffType]}`
+            );
+            addNodeArrayPropertyDiffs(key, arrValue, diffType, typedDiff);
+          });
+      } else {
+        const diffType = getDiffType(value);
+        ramlToolLogger.debug(
+          `Adding difference of node: ${typedDiff.id}, property: ${key}, difference type: ${DiffType[diffType]}`
+        );
+        addNodePropertyDiffs(key, value, diffType, typedDiff);
+      }
+    });
+  if (_.isEmpty(typedDiff.added) && _.isEmpty(typedDiff.removed)) {
+    //when there are just moves and no actual changes, ignore the typed diff
+    typedDiff = undefined;
+  }
+  return typedDiff;
+}
+
+/**
+ * Get the type of the difference from the difference array generated by jsondiffpatch
+ * @param diff Difference of a node or a node property
+ *
+ * @return DiffType Type of the difference
+ */
+function getDiffType(diff: Array<any>): DiffType {
+  if (diff.length === 1) {
```

Comment:
switch would look nice here

Reply:
since there is another if when the length is 3, I feel this looks cleaner. leaving it for now

Reply:
We can have another private method for those sub conditionals..Its not a big deal though..

Reply:
just feel its easy to read this way.. we can change it later .. not a big deal for me too :)